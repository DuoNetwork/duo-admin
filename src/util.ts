import moment from 'moment';
import request from 'request';
import * as CST from './constants';
import { IOption } from './types';

class Util {
	public log(text: any): void {
		console.log(moment().format('HH:mm:ss.SSS') + ' ' + text);
	}

	// public truncateNum(num: number, decimal: number = 3): number {
	// 	let x = num.toString();
	// 	x = x.slice(0, x.indexOf('.') + decimal);
	// 	return Number(x);
	// }

	public get(url: string, headerOthers?: object): Promise<any> {
		return new Promise((resolve, reject) =>
			request(
				{
					url,
					headers: Object.assign(
						{
							'user-agent': 'node.js'
						},
						headerOthers || {}

					)
				},
				(error, res, body) => {
					if (error) reject(error);
					else if (res.statusCode === 200) resolve(body);
					else reject('Error status ' + res.statusCode + ' ' + res.statusMessage);
				}
			)
		);
	}

	public generateRandomIdx(max: number, alpha: number): number[] {
		const output: number[] = [];
		if (alpha === 1) for (let i = 0; i < max; i++) output.push(i);
		else {
			const num = Math.floor(max * alpha);

			while (output.length < num) {
				const item = Math.floor(Math.random() * max);
				if (output.indexOf(item) < 0) output.push(item);
			}
		}

		return output;
	}

	public postJson(url: string, json: object): Promise<object> {
		return new Promise((resolve, reject) =>
			request(
				{
					url,
					method: 'POST',
					headers: {
						'Content-Type': 'application/json'
					},
					json
				},
				(error, res, body) => {
					if (error) reject(error);
					else if (res.statusCode === 200) resolve(body);
					else reject('Error status ' + res.statusCode + ' ' + res.statusMessage);
				}
			)
		);
	}

	public parseOptions(argv: string[]): IOption {
		const option = {
			live: process.argv.includes('live'),
			dynamo: process.argv.includes('dynamo'),
			aws: process.argv.includes('aws'),
			gcp: process.argv.includes('gcp'),
			azure: process.argv.includes('azure'),
			force: process.argv.includes('force'),
			gasPrice: 5e9,
			gasLimit: 200000,
			eth: 0,
			address: '',
			privateKey: '',
			price: 0,
			source: '',
			pwd: '',
			event: '',
			provider: '',
			contractState: '',
			accountNum: 1,
			saveAccount: false,
			from: '',
			to: '',
			value: 0,
			index: 0,
			total: 10,
			minEther: 0.02,
			alpha: 0.3,
			amtA: 0,
			amtB: 0,
			generator: ''
		};
		for (let i = 3; i < argv.length; i++) {
			const args = argv[i].split('=');
			switch (args[0]) {
				case 'gasPrice':
					option.gasPrice = Number(args[1]) || option.gasPrice;
					break;
				case 'gasLimit':
					option.gasLimit = Number(args[1]) || option.gasLimit;
					break;
				case 'eth':
					option.eth = Number(args[1]) || option.eth;
					break;
				case 'address':
					option.address = args[1] || option.address;
					break;
				case 'privateKey':
					option.privateKey = args[1] || option.privateKey;
					break;
				case 'price':
					option.price = Number(args[1]) || option.price;
					break;
				case 'pwd':
					option.pwd = args[1] || option.pwd;
					break;
				case 'source':
					option.source = args[1] || option.source;
					break;
				case 'event':
					option.event = args[1] || option.event;
					break;
				case 'provider':
					option.provider = args[1] || option.provider;
					break;
				case 'contractState':
					option.contractState = args[1] || option.contractState;
					break;
				case 'accountNum':
					option.accountNum = Number(args[1]) || option.accountNum;
					break;
				case 'saveAccount':
					option.saveAccount = args[1] === 'yes' || option.saveAccount;
					break;
				case 'from':
					option.from = args[1] || option.from;
					break;
				case 'total':
					option.total = Number(args[1]) || option.total;
					break;
				case 'minEther':
					option.minEther = Number(args[1]) || option.minEther;
					break;
				case 'alpha':
					option.alpha = Number(args[1]) || option.alpha;
					break;
				case 'amtA':
					option.amtA = Number(args[1]) || option.amtA;
					break;
				case 'amtB':
					option.amtB = Number(args[1]) || option.amtB;
					break;
				case 'value':
					option.value = Number(args[1]) || option.value;
					break;
				case 'index':
					option.index = Number(args[1]) || option.index;
					break;
				case 'generator':
					option.generator = args[1] || option.generator;
					break;
				default:
					break;
			}
		}

		if (!option.provider)
			if (option.source === CST.SRC_MYETHER)
				option.provider = option.live
					? CST.PROVIDER_MYETHER_LIVE
					: CST.PROVIDER_MYETHER_DEV;
			else if (option.source === CST.SRC_INFURA)
				option.provider = option.live ? CST.PROVIDER_INFURA_LIVE : CST.PROVIDER_INFURA_DEV;
			else {
				option.provider = CST.PROVIDER_LOCAL_WS;
				option.source = '';
			}

		return option;
	}

	public getDynamoRole(tool: string, useDynamo: boolean): string {
		switch (tool) {
			case 'bitfinex':
			case 'gemini':
			case 'kraken':
			case 'gdax':
				return useDynamo ? CST.AWS_DYNAMO_ROLE_TRADE : CST.AWS_DYNAMO_ROLE_STATUS;
			case 'commit':
			case 'node':
				return CST.AWS_DYNAMO_ROLE_STATUS;
			case 'subscribe':
				return CST.AWS_DYNAMO_ROLE_EVENT;
			case 'hourly':
				return CST.AWS_DYNAMO_ROLE_HOURLY;
			case 'minutely':
				return CST.AWS_DYNAMO_ROLE_MINUTELY;
			default:
				return '';
		}
	}

	public getStatusProcess(tool: string, option: IOption) {
		let type = '';
		const platform = option.azure ? '_AZURE' : option.gcp ? '_GCP' : '_AWS';
		const privacy = option.dynamo ? '_PUBLIC' : '_PRIVATE';
		let source = '';

		switch (tool) {
			case 'bitfinex':
			case 'gemini':
			case 'kraken':
			case 'gdax':
				type = 'PRICE';
				source = '_' + tool.toUpperCase();
				break;
			case 'subscribe':
				type = 'EVENT';
				source =
					'_' +
					([CST.EVENT_START_PRE_RESET, CST.EVENT_START_RESET].includes(option.event)
						? option.event.toUpperCase()
						: 'OTHERS');
				break;
			case 'commit':
				type = 'FEED';
				break;
			case 'hourly':
				type = 'HOURLY';
				break;
			case 'minutely':
				type = 'MINUTELY';
				break;
			case 'node':
				type = 'CHAIN';
				break;
			default:
				return '';
		}

		return type + platform + privacy + source;
	}

	public getNowTimestamp() {
		return moment().valueOf();
	}

	public isNumber(input: any): boolean {
		const num = Number(input);
		return isFinite(num) && !isNaN(num);
	}

	public isEmptyObject(obj: object | undefined | null): boolean {
		if (!obj) return true;

		for (const prop in obj) if (obj.hasOwnProperty(prop)) return false;

		return true;
	}
}

const util = new Util();
export default util;
