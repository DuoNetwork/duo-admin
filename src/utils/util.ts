import moment from 'moment';
import * as CST from '../common/constants';
import { IOption } from '../common/types';
// import { Options } from 'aws-cli-js';

class Util {
	public logLevel: string = CST.LOG_INFO;

	private log(text: any, level?: string): void {
		if (level && CST.LOG_RANKING[this.logLevel] >= CST.LOG_RANKING[level])
			console.log(`${moment().format('HH:mm:ss.SSS')} [${level}]: ` + text);
	}

	public logInfo(text: any): void {
		this.log(text, CST.LOG_INFO);
	}

	public logDebug(text: any): void {
		this.log(text, CST.LOG_DEBUG);
	}

	public logError(text: any): void {
		this.log(text, CST.LOG_ERROR);
	}

	public composeQuery(paras: { [key: string]: any }): string {
		let query: string = '?';
		for (const key in paras) query += key + '=' + paras[key] + '&';
		query = query.slice(0, -1);
		return query;
	}

	public parseOptions(argv: string[]): IOption {
		const option = {
			forceREST: argv.includes('rest'),
			live: process.argv.includes('live'),
			dbLive: process.argv.includes('dbLive'),
			server: process.argv.includes('server'),
			dynamo: process.argv.includes('dynamo'),
			aws: process.argv.includes('aws'),
			gcp: process.argv.includes('gcp'),
			azure: process.argv.includes('azure'),
			force: process.argv.includes('force'),
			pair: '',
			contractType: 'Beethoven',
			tenor: 'Perpetual',
			assets: [''],
			sources: [''],
			exSources: [''],
			gasPrice: 5e9,
			gasLimit: 200000,
			source: '',
			event: '',
			provider: '',
			period: 1,
			base: 'USD',
			quote: 'ETH'
		};
		for (let i = 3; i < argv.length; i++) {
			const args = argv[i].split('=');
			switch (args[0]) {
				case 'exSources':
					option.exSources = args[1].split(',');
					break;
				case 'sources':
					option.sources = args[1].split(',');
					break;
				case 'pair':
					option.pair = args[1].replace('_', '|') || option.pair;
					break;
				case 'assets':
					option.assets = args[1].split(',');
					break;
				case 'gasPrice':
					option.gasPrice = Number(args[1]) || option.gasPrice;
					break;
				case 'gasLimit':
					option.gasLimit = Number(args[1]) || option.gasLimit;
					break;
				case 'source':
					option.source = args[1] || option.source;
					break;
				case 'event':
					option.event = args[1] || option.event;
					break;
				case 'provider':
					option.provider = args[1] || option.provider;
					break;
				case 'period':
					option.period = Number(args[1]);
					break;
				case 'base':
					option.base = args[1] || option.base;
					break;
				case 'quote':
					option.quote = args[1] || option.quote;
					break;
				case 'contractType':
					option.contractType = args[1] || option.contractType;
					break;
				case 'tenor':
					option.tenor = args[1] || option.tenor;
					break;
				default:
					break;
			}
		}

		return option;
	}

	public getPeriodStartTimestamp(timestamp: number, period: number = 1) {
		// const now = this.getUTCNowTimestamp();
		return Math.floor(timestamp / 60000 / period - 1) * 60000 * period;
	}

	public timestampToString(ts: number) {
		//just for debugging purpose
		return moment.utc(ts).format('YYYY-MM-DDTHH:mm:ss');
	}

	public getStatusProcess(tool: string, option: IOption) {
		let type = '';
		const platform = option.azure ? '_AZURE' : option.gcp ? '_GCP' : '_AWS';
		console.log('dynamo', option.dynamo);
		const privacy = option.dynamo ? '_PUBLIC' : '_PRIVATE';
		let source = '';

		switch (tool) {
			case CST.TRADES:
				type = 'TRADE';
				source = option.source ? '_' + option.source.toUpperCase() : '';
				break;
			case CST.TRIGGER:
			case CST.FETCH_EVENTS:
				type = 'EVENT';
				source =
					'_' +
					([CST.EVENT_START_PRE_RESET, CST.EVENT_START_RESET].includes(option.event)
						? option.event.toUpperCase()
						: 'OTHERS');
				break;
			case CST.COMMIT:
				type = 'FEED';
				break;
			case CST.CLEAN_DB:
				type = 'CLEANDB';
				break;
			case CST.DB_PRICES:
				type = 'PRICE';
				if (option.period === 1) source = '_MINUTELY';
				else if (option.period === 60) source = '_HOURLY';
				else source = '';
				break;
			case 'minutely':
				type = 'MINUTELY';
				break;
			case CST.FETCH_PRICE:
				type = 'FETCH';
				break;
			default:
				return '';
		}

		return type + platform + privacy + source;
	}

	public getUTCNowTimestamp() {
		return moment().valueOf();
	}

	public isNumber(input: any): boolean {
		const num = Number(input);
		return isFinite(num) && !isNaN(num);
	}

	public isEmptyObject(obj: object | undefined | null): boolean {
		if (!obj) return true;

		for (const prop in obj) if (obj.hasOwnProperty(prop)) return false;

		return true;
	}

	public async sleep(ms: number) {
		return new Promise(resolve => {
			setTimeout(resolve, ms);
		});
	}
}

const util = new Util();
export default util;
