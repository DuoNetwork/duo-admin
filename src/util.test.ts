import util from './util';

test('isNumber() return true for numbers', () => {
	expect(util.isNumber(5)).toBe(true);
	expect(util.isNumber(5.0)).toBe(true);
});

test('isNumber() return true for empty string and null', () => {
	expect(util.isNumber('')).toBe(true);
	expect(util.isNumber(null)).toBe(true);
});

test('isNumber() return true for number strings', () => {
	expect(util.isNumber('5')).toBe(true);
	expect(util.isNumber('5.0')).toBe(true);
});

test('isNumber() return false for other strings', () => {
	expect(util.isNumber('5.0s')).toBe(false);
	expect(util.isNumber('test')).toBe(false);
	expect(util.isNumber('NaN')).toBe(false);
});

test('isNumber() return false for undefined, infinity, NaN', () => {
	expect(util.isNumber(undefined)).toBe(false);
	expect(util.isNumber(Infinity)).toBe(false);
	expect(util.isNumber(NaN)).toBe(false);
});
