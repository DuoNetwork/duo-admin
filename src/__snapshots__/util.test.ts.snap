// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseOptions 1`] = `
Object {
  "accountNum": 1,
  "addr1": "",
  "addr2": "",
  "address": "",
  "alpha": 0.3,
  "amtA": 0,
  "amtB": 0,
  "assets": Array [
    "quote",
    "base",
  ],
  "aws": false,
  "azure": false,
  "contractState": "",
  "dbLive": false,
  "dynamo": false,
  "endBlk": 0,
  "eth": 0,
  "event": "",
  "exSources": Array [
    "",
  ],
  "force": false,
  "forceREST": false,
  "from": "",
  "gasLimit": 200000,
  "gasPrice": 5000000000,
  "gcp": false,
  "index": 0,
  "key": "",
  "live": false,
  "minEther": 0.02,
  "numOfHours": 2,
  "numOfMinutes": 2,
  "pair": "ETH|USD",
  "period": 10,
  "price": 0,
  "privateKey": "",
  "provider": "ws://localhost:8546",
  "pwd": "",
  "saveAccount": false,
  "server": false,
  "source": "source",
  "sources": Array [
    "",
  ],
  "to": "",
  "total": 10,
  "value": 0,
}
`;
